File: 17_binarytreeleaves.py
class TreeNode:
    def __init__(self, value=0, left=None, right=None):
        self.value = value
        self.left = left
        self.right = right

def count_leaves(root):
    """
    Подсчитывает количество листьев в бинарном дереве (рекурсивно).
    
    Лист - это узел, у которого нет потомков (left и right равны None).
    
    Args:
        root: корень бинарного дерева
    
    Returns:
        int: количество листьев
    """
    # Базовый случай: пустое дерево
    if root is None:
        return 0
    
    # Если узел является листом (нет потомков)
    if root.left is None and root.right is None:
        return 1
    
    # Рекурсивно считаем листья в левом и правом поддеревьях
    return count_leaves(root.left) + count_leaves(root.right)

# Создаём дерево:
#       1
#      / \
#     2   3
#    / \   \
#   4   5   6
#  /
# 7

root = TreeNode(1)
root.left = TreeNode(2)
root.right = TreeNode(3)
root.left.left = TreeNode(4)
root.left.right = TreeNode(5)
root.right.right = TreeNode(6)
root.left.left.left = TreeNode(7)

# Листья: 7, 5, 6 (всего 3)

print(f"Кол-во листьев: {count_leaves(root)}")  # 3

File: 17_binarytreeleaves.cpp
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    
    TreeNode(int value) : val(value), left(nullptr), right(nullptr) {}
};

class BinaryTree {
public:
    // Рекурсивный метод подсчета листьев
    int countLeaves(TreeNode* root) {
        // Базовый случай: пустое дерево
        if (root == nullptr) {
            return 0;
        }
        
        // Если узел является листом (нет детей)
        if (root->left == nullptr && root->right == nullptr) {
            return 1;
        }
        
        // Рекурсивно считаем листья в левом и правом поддереве
        return countLeaves(root->left) + countLeaves(root->right);
    }
};

int main() {
    // Создаем дерево:
    //       1
    //      / \
    //     2   3
    //    / \
    //   4   5
    
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->right = new TreeNode(5);
    
    BinaryTree tree;
    cout << "Количество листьев (рекурсия): " << tree.countLeaves(root) << endl;
    // Вывод: 3
    
    return 0;
}
